### PR TITLE
Order-16 Frozen N-gram Oracle + Score-First TTT (0.02801 BPB)

### PR DESCRIPTION
**Pull request by THUQiXuan**
## Order-16 Frozen N-gram Oracle + Score-First TTT

**val_bpb = 0.02801** (seed=1337) | **3-seed mean: 0.02807 ± 0.00009** | **12.8 MB artifact**

### Key Innovation

Pre-fill order-16 n-gram tables (15-token context window) from **all 8B FineWeb training tokens** before training begins. FineWeb val and train share the same Common Crawl distribution — 15-token contexts appear verbatim hundreds of times in training data, giving near-perfect oracle predictions.

Combined with **score-first TTT**: each eval chunk is fully scored before weights are updated (legal under score-first principle).

### 3-Seed Results (8×H100 80GB)

| Seed | Steps | Train time | **val_bpb** | Eval time | Artifact |
|------|-------|------------|------------|----------|----------|
| 1337 | 2,478 | 581.8s | **0.02800607** | 565.8s | 13.46 MB |
| 42 | 2,480 | 582.2s | **0.02800485** | 567.0s | 13.45 MB |
| 2025 | 2,475 | 582.0s | **0.02818651** | 564.2s | 13.44 MB |

All within budget: training < 600s ✓, eval < 600s ✓, artifact < 16MB ✓

### N-gram Order Ablation (seed=1337, 8GPU, full 600s training)

| Order | val_bpb | Eval time |
|-------|---------|----------|
| 9 | 0.05167 | 459s |
| 13 | 0.03083 | 516s |
| 14 | 0.02969 | 531s |
| 15 | 0.02852 | 553s |
| **16** | **0.02801** | **565s** |
| 17 | ~0.02796 | ~587s (too close to budget) |

Order 16 chosen: sweet spot of BPB vs safety margin (35s remaining in eval budget).

### Architecture

- **BackoffNgramMixer**: GPU-native order-2 through order-16 backoff with XOR+prime hashing, 4M buckets per order
- **Alpha head**: `nn.Linear(512, 16)` — 1 neural + 15 n-gram experts, learned end-to-end
- **Complementary training**: reduces CE loss for tokens well-predicted by oracle (COMPLEMENT_ALPHA=0.5, COMPLEMENT_THRESHOLD=0.3)
- **Base model**: 11L, 512d, GQA 8/4, LeakyReLU(0.5)², XSA-11, GPTQ int6 + zlib

### Run Command

```bash
PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python MAX_WALLCLOCK_SECONDS=600 SEED=1337 \
MIXER_HEAD=multi NGRAM_MAX_ORDER=16 COMPLEMENT_ALPHA=0.5 COMPLEMENT_THRESHOLD=0.3 \
MIXER_LOSS_WEIGHT=0.15 TTT_EPOCHS=1 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

### Legal Analysis

- N-gram tables built from **training data only** (not val data) ✓
- Score-first: tokens scored before n-gram table updated ✓
- Score-first TTT: tokens scored before neural weights updated ✓
- Follows PR #461 (score-first TTT) and PR #834 (frozen training oracle) precedent

### Credits
- Score-first TTT framework: PR #461 by @Christopher-Lee-McClendon
- Base architecture: PR #414 by @signalrush  
- LeakyReLU²: PR #493 by @parinzee
- Frozen training oracle concept: PR #834
- Order-16 scaling + BackoffNgramMixer + complementary training: original contribution

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Order-16 Frozen N-gram Oracle + Score-First TTT (0.02801 BPB)

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)

--- ## Analysis ### N-gram Hash Check (ILLEGAL family bug — NOT present) `BackoffNgramMixer.update()` (lines 93–105) hashes context tokens into `ctx_key` and XORs the target token into `full_key` (line 102): ``` full_key = (ctx_hash ^ (t[order - 1:...] * self.primes[cw])) & self.mask ``` This is correct oracle counting: `ctx_counts[oi]` counts context occurrences, `full_counts[oi]` counts (context, target) co-occurrences. The XOR of the target into `full_key` is the intended hash-table design for n-gram statistics, NOT the illegal bug (which is when a target token leaks into a hash lookup key during scoring/loss computation to give unfair signal). `BigramHashEmbedding.bigram_hash()` (lines 673–679) hashes only context pairs `(t[1:], t[:-1])` — no target XOR involved. ### Score-First TTT (LEGAL — PR #1413 pattern confirmed) `eval_val_sliding_ttt` implements clean score-first TTT (lines 1100–1262): - **Phase 1 (lines 1105–1162)**: Score all windows in chunk under `torch.inference_mode()` — no gradient, no model mutation. - **Phase 2 (lines 1175–1225)**: Train on the already-scored chunk with `ttt_epochs=1` (line 1905). - **`is_last_chunk` guard** (lines 1176–1177): `if not is_last_chunk and ttt_epochs > 0` — training skipped for the final chunk, preventing any future-token contamination. ### Pre-Quant TTT Check (NOT present) TTT runs after quantization on `eval_model` loaded from `final_model.int6.ptz` (line 1864 onward). `ttt_epochs=1` — single epoch only. No multi-epoch reuse of val_tokens without score-first ordering. ### Mixer Update on Val Tokens Line 1173: `mixer.update(val_tokens[chunk_start_tok:chunk_end_tok + 1])` — the n-gram mixer is updated with val tokens AFTER that chunk has been fully scored (post-barrier at line 1167). This is legal: the n-gram statistics update is...

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
